#Imbalanced Classification

## 1. Defining the Question

predict whether a customer will leave the bank soon using past behaviour data

### b) Defining the Metric for Success

Build a model with the maximum possible F1 score. on whether the customer is about to leave

Beta Bank customers are leaving: little by little, chipping away every month. The bankers
figured out it’s cheaper to save the existing customers rather than to attract new ones.
We need to predict whether a customer will leave the bank soon. You have the data on
clients’ past behavior and termination of contracts with the bank.


### d) Recording the Experimental Design

Describe the steps/approach that you will use to answer the given question.




1.   Data Exploration

1.  Data Preparation

1.   Data Modeling
2.   Summary of Findings and Recommendations

2.   Imbalanced Classification



















### e) Data Relevance





Data provided is sufficient to preict if the cutomer will leave

## 2. Reading the Data

In [149]:
# Importing our libraries 
# ---
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder, MinMaxScaler, StandardScaler
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split 
from sklearn.metrics import f1_score
from sklearn.metrics import roc_curve
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

In [150]:
# Load the data below
# --- 
# Dataset url = : : https://bit.ly/2XZK7Bo

# --- 
df = pd.read_csv('https://bit.ly/2XZK7Bo')

In [151]:
#first and last rows
#
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [152]:

df.tail()

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
9995     5.0       0.00              2          1               0   
9996    10.0   57369.61              1          1               1   
9997     7.0       0.00              1          0               1   
9998     3.0   75075.31              2          1               0   
9999     NaN  130142.79              1          1               0   

      EstimatedSalary  Exited  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0

Record your observations below:


*   The data provided has 10000 rows and 14 columns



## 4. Data Preparation

### Performing Data Cleaning

In [153]:
# Checking missing entries of all the variables. 
# ---
# df.isnull().sum()
pd.DataFrame(df.isnull().sum(), columns=["Number of Missing Values"])

Number of Missing Values
RowNumber                               0
CustomerId                              0
Surname                                 0
CreditScore                             0
Geography                               0
Gender                                  0
Age                                     0
Tenure                                909
Balance                                 0
NumOfProducts                           0
HasCrCard                               0
IsActiveMember                          0
EstimatedSalary                         0
Exited                                  0

We observe the following from our dataset:

*  Column Teunre has 909 missing values 



In [154]:

df.columns = df.columns.str.lower()
df.head()

rownumber  customerid   surname  creditscore geography  gender  age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   tenure    balance  numofproducts  hascrcard  isactivemember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   estimatedsalary  exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

We observe the following from our dataset:

*   We renamed all columns to have lower cases and checked to confirm the changes. All columns now have lower cases in the column names.



In [155]:

df.duplicated().sum()

0

We observe the following from our dataset:

*   There are no duplicates in our data



In [156]:
# Checking if any of the rows are all null

sum(df.isnull().all(axis = 1))

0

We observe the following from our dataset:

*   No row contains completely null values



In [157]:


df_churn = df.copy()
df_churn.head()

rownumber  customerid   surname  creditscore geography  gender  age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   tenure    balance  numofproducts  hascrcard  isactivemember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   estimatedsalary  exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

fixing Missing values  in tenure

In [158]:
df_churn['tenure'] = df_churn['tenure'].fillna(0)
pd.DataFrame(df_churn.isnull().sum(), columns=["Number of Missing Values"])

Number of Missing Values
rownumber                               0
customerid                              0
surname                                 0
creditscore                             0
geography                               0
gender                                  0
age                                     0
tenure                                  0
balance                                 0
numofproducts                           0
hascrcard                               0
isactivemember                          0
estimatedsalary                         0
exited                                  0

In [159]:
# Drop and reassign dataframe
df_churn.drop(columns=['rownumber', 'customerid', 'surname','geography','gender'], inplace=True)

# View first few 5 rows
df_churn.head()

creditscore  age  tenure    balance  numofproducts  hascrcard  \
0          619   42     2.0       0.00              1          1   
1          608   41     1.0   83807.86              1          0   
2          502   42     8.0  159660.80              3          1   
3          699   39     1.0       0.00              2          0   
4          850   43     2.0  125510.82              1          1   

   isactivemember  estimatedsalary  exited  
0               1        101348.88       1  
1               1        112542.58       0  
2               0        113931.57       1  
3               0         93826.63       0  
4               1         79084.10       0

Train dataset

In [160]:

df_churn,df_valid = train_test_split(df_churn, test_size=0.35, random_state=1234)
print(df_churn.shape)
print(df_valid.shape)

(6500, 9)
(3500, 9)


In [161]:
#create features and target for both train and test
f_train = df_churn.drop(columns=['exited'])
t_train = df_churn['exited']
f_test = df_valid.drop(columns=['exited'])
t_test = df_valid['exited']


In [162]:

DT_model = DecisionTreeClassifier(random_state=1234, max_depth=2)
DT_model.fit(f_train, t_train)  

print('Decision tree with depth  of 2 =', tree_model.score(f_train, t_train))

DT_model = DecisionTreeClassifier(random_state=1234, max_depth=3)
DT_model.fit(f_train, t_train)  

print('Decision tree with depth  of 3 =', tree_model.score(f_train, t_train))

Decision tree with depth  of 2 = 0.8407692307692308
Decision tree with depth  of 3 = 0.8407692307692308


In [163]:

RF_model = RandomForestClassifier(class_weight='balanced' ,random_state=1234, n_estimators=2)
RF_model.fit(f_train, t_train)
print('Random forest with n_estimators 2 ', RF_model.score(f_train, t_train))
RF_model = RandomForestClassifier(class_weight='balanced' ,random_state=1234, n_estimators=3)
RF_model.fit(f_train, t_train)
print('Random forest with n_estimators 3 ', RF_model.score(f_train, t_train))

Random forest with n_estimators 2  0.922
Random forest with n_estimators 3  0.9635384615384616


F1

In [164]:


RF = RandomForestClassifier(n_estimators = 100, random_state = 40)

RF.fit(f_train,t_train)


RF_predict = RF.predict(f_valid)

RF_score = accuracy_score(RF_predict, t_valid)

print('F1 score is ',f1_score(t_valid,rf_predict))


F1 score is  0.5554553651938683


Random Forest with estimator 3 gives the highest prediction of 0.96